utils/throttler.go

```diff
@@ -0,0 +1,51 @@
+package utils
+
+import (
+	"errors"
+	"math"
+	"sync/atomic"
+)
+
+var ErrResourceBusy = errors.New("resource busy, try again")
+
+type Throttler[T any] struct {
+	resource *T
+	sem      chan struct{}
+	queue    atomic.Int32
+
+	maxQueueLen int32
```

Comment:
Do you think `Unit32` would be more appropriate?

Reply:
Yeah, I don't like that I had to use signed types there either. I wanted to keep `queue` and `maxQueueLen` same types to avoid running in to weird issues when comparing them due to casting one of them to the other type. And `queue` is a signed type to be able to do `t.queue.Add(-1)`.